python/mxnet/gluon/contrib/estimator/event_handler.py

```diff
@@ -0,0 +1,459 @@
+                name, value = monitor.get()
+                msg += '%s : %.4f ' % (name, value)
+            self.logger.info(msg)
+            self.current_epoch += 1
+            self.batch_index = 0
+
+
+class CheckpointHandler(BatchEnd, EpochEnd):
+    """Save the model after every epoch.
+
+    :py:class:`CheckpointHandler` save the network parameters every epoch
+
+    Parameters
+    ----------
+    filepath : str
+        file name to save the parameters, it can contain directories,
+        for example: ./saved_model/resnet.params
+    monitor: EvalMetric
+        the metrics to monitor
+    verbose: int, default 0
+        verbosity mode
+    save_best_only: bool
+        if True, only save the parameters if monitored value improved
+    mode: str, default 'auto'
+        one of {auto, min, max}, if `save_best_only=True`, the comparison to make
+        and determine if the monitored value has improved
+    period: int, default 1
+        intervals between saving the network
+    """
+
+    def __init__(self,
+                 filepath,
+                 monitor=None,
+                 verbose=0,
+                 save_best_only=False,
+                 mode='auto',
+                 epoch_period=1,
+                 batch_period=None):
+        self.monitor = monitor
+        self.verbose = verbose
+        self.filepath = filepath
+        self.save_best_only = save_best_only
+        if self.save_best_only and not isinstance(self.monitor, EvalMetric):
+            raise ValueError("To save best model only, please provide one of the metric objects as monitor, "
+                             "You can get these objects using estimator.prepare_loss_and_metric()")
+        self.epoch_period = epoch_period
+        self.batch_period = batch_period
+        self.num_batches = 0
+        self.num_epochs = 0
+        self.logger = logging.getLogger(__name__)
+
+        if mode not in ['auto', 'min', 'max']:
+            warnings.warn('ModelCheckpoint mode %s is unknown, '
+                          'fallback to auto mode.' % (mode),
+                          RuntimeWarning)
+            mode = 'auto'
+
+        if mode == 'min':
+            self.monitor_op = np.less
+            self.best = np.Inf
+        elif mode == 'max':
+            self.monitor_op = np.greater
+            self.best = -np.Inf
+        else:
+            # use greater for accuracy and less otherwise
+            if 'acc' in self.monitor.get()[0].lower():
+                self.monitor_op = np.greater
+                self.best = -np.Inf
+            else:
+                self.monitor_op = np.less
+                self.best = np.Inf
+
+    def batch_end(self, estimator, *args, **kwargs):
+        self._save_checkpoint(estimator.net, "Batch", self.num_batches)
+        self.num_batches += 1
+
+    def epoch_end(self, estimator, *args, **kwargs):
+        self._save_checkpoint(estimator.net, "Epoch", self.num_epochs)
```

Comment:
this will create a lot of files with unique names, should we also allow to save the last only in the same filename?

Reply:
I'm using save parameter only, epcoh/batch number is not saved into file name.  num_epoch and num_batch is passed only to check if the saving frequency has reached.